Add Prettier formatting script

jsnajdr opened the pull request:
This PR adds support for [wp-prettier](https://github.com/automattic/wp-prettier), a Prettier fork that follows the WordPress code conventions and inserts extra spaces inside parens:
```js
function fun( param ) {
  foo( [ 1, 2 ], ( val ) => {
    return `key-${ val }`;
  } );
}
```

I've been maintaining the fork for 2+ years, from version 1.5 to the latest 1.18.2, and it's been super-useful in Automattic projects like Calypso or Jetpack.

**Add a `format-js` script to `@wordpress/scripts`:**
There is a new script that can be used to format a source directory like this:
```
wp-scripts format-js packages/compose
```
The arguments are a list of files and directories to be formatted (globs are possible, too), and it will format all matching JS files, while respecting the ignores in the default `.eslintignore`.

**Remove conflicting ESLint rules:**
Some ESLint rules that guard formatting are known to conflict with Prettier (difference in opinions about what's the "right" format), so I'm disabling these that check for things that are automatically handled by Prettier. There's also an [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) ESLint plugin for that.

**Format `packages/compose` to get a taste of Prettier:**
This PR also includes formatting of one small module to get a taste of how the formatting changes look like.

**TODO: incremental adoption**
One proven way how to adopt Prettier incrementally in a large codebase is this one:
1. Whenever a file is formatted with Prettier, add a `/** @format */` pragma at the beginning. That tells the automated tooling to try keep it formatted.
2. In a pre-commit hook, check for `@format` pragma in the modified file and reformat it before committing if the pragma is present. Can be done with `lint-staged` tool that we already use.

The tooling for that is NOT in this PR. Can be added later if there's an interest in Prettier in the first place.

jsnajdr committed:
Scripts: export the fromProjectRoot function from scripts/utils module

ntwb commented on packages/scripts/scripts/format-js.js:
```diff
@@ -0,0 +1,76 @@
+	hasProjectFile( '.prettierrc.yaml' ) ||
+	hasProjectFile( '.prettierrc.yml' ) ||
+	hasProjectFile( '.prettierrc.js' ) ||
+	hasProjectFile( '.prettierrc.config.js' ) ||
```
```suggestion
	hasProjectFile( 'prettierrc.config.js' ) ||
```

No leading `.` period, Prettier uses cosmiconfig, see https://prettier.io/docs/en/configuration.html

ntwb commented on packages/scripts/package.json:
```diff
@@ -44,13 +44,15 @@
 		"lodash": "^4.17.15",
 		"minimist": "^1.2.0",
 		"npm-package-json-lint": "^3.6.0",
+		"prettier": "npm:wp-prettier@^1.18.2",
```
What functionality does the `npm:` prefix here perform if you don't mind, I've not seen this?

swissspidy replied:
It aliases `prettier` to `wp-prettier`, so it will actually use the `wp-prettier` package under the hood. This was introduced in npm 6.9.

ntwb replied:
Neat, thanks, this is pretty cool

gziolo replied:
We would need to enforce that `check-engines` requires npm 6.9 or higher by default.

swissspidy replied:
Even cooler would be not having to maintain such a fork, but progress is slow on that front: https://github.com/prettier/prettier/issues/5919

jsnajdr replied:
Note that even if prettier/prettier#5919 was fixed and we could ship `wp-prettier` as a plugin, it wouldn't change much. We'd still need to maintain a fork of the JS formatter, and then ship the forked formatter as a plugin that replaces the original JS formatter. The code and the work is still the same, just shipped in different packaging.

I could fix prettier/prettier#5919 myself if needed, but with package aliasing support shipping both in NPM and Yarn, the aliasing method works equally well.

jsnajdr replied:
> This was introduced in npm 6.9.

Yes, and it's the main reason why I waited so long before proposing to use the fork. Before package aliases, it could be distributed only as a tarball URL or a GitHub tag ref, which is kind of ugly and doesn't support versioning well. Now, distributing forks of things is much cleaner and easier.

ntwb commented on packages/scripts/config/.prettierrc.js:
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+	useTabs: true,
+	tabWidth: 2,
```
Should this be defined? Some devs like 4 spaces, some 2, even 8.

Leaving this as a developer preference for them to set in their Editor/IDE is in line with how WordPress Core uses `.editorconfig` by not defining a _tab width_

ntwb replied:
Thanks for the explanation 👍🏼

ntwb commented:
Thanks for the PR @jsnajdr I've added a couple of comments for starters...

What I would like to see though is how Prettier has, or if it has since we looked at it in #4628, our conclusion at the time was this (copying here for ease ref) https://github.com/WordPress/gutenberg/pull/4628#issuecomment-360071509:
> Prettier and WordPress are both pretty opinionated, though there's only a couple of instances where those opinions really opposed each other, thankfully ESLint's `--fix` was able to bring a satisfactory compromise. 
> 
> To that end examing whats left after WordPress has applied its ESLint fixes to the changes made by Prettier is relatively minor, primarily these changes relate to _line length_, where Prettier refactors the code to not exceed `x` characters in length, it does a pretty good job of that for the most part. As noted in the comments above sometimes it's not ideal where and when it does this it makes for poorly formatted code.
> 
> For the remaining instances of stylistic formatting that Prettier made we can look to add ESLint rules to detect these style inconsistencies and update our coding standards to match. Hopefully we can contribute ESLint rules and fixes upstream as much as possible for the benefit of all ESLint users.
> 
> I don't think Prettier is the right solution for WordPress Core, we'll continue to enhance and iterate our JavaScript Coding Standards with ESLint.
> 
> Thanks to everyone who has tested and explored the world of Prettier with us here, I've had a blast testing this, closing this PR now, thanks again everyone 😄 

I would love to see how Prettier fares today with some of the issues we raised as concerns:

- [ ] Line Length https://github.com/WordPress/gutenberg/pull/4628#issuecomment-359604670
- [ ] Inline docs https://github.com/WordPress/gutenberg/pull/4628#issuecomment-359607250
- [ ] Indentation https://github.com/WordPress/gutenberg/pull/4628#issuecomment-359612556
- [ ] Removing unnecessary brackets https://github.com/WordPress/gutenberg/pull/4628#issuecomment-359657339 https://github.com/WordPress/gutenberg/pull/4628#issuecomment-359736128
- [ ] Removing brackets https://github.com/WordPress/gutenberg/pull/4628#issuecomment-359657339

gziolo commented:
Noting that this PR addresses #2819 which was closed a long time ago due to the issues listed in the previous comment by @ntwb.

jsnajdr commented:
> I would love to see how Prettier fares today with some of the issues we raised as concerns:

To be honest, the formatting hasn't changed much 🙂 Looking at the concerns that had been raised, I see two where a fix could be proposed upstream:

**Botching up alignment of docblock comments inside JSX**
```jsx
<div>
  { /**
   * i'm a docblock
   */ }
</div>;
```
This is clearly a Prettier bug, as it happens only when the JSX expression container is otherwise empty. Adding anything to it fixes the comment formatting.

**Unpleasant breaking of long curry chains with `arrowParens=always`**
```js
const f = ( a ) => ( b ) => (
  c
) => find( a, b, c );
```
The `arrowParens=avoid` version of this code doesn't break the line, even if the curry chain is longer than the maximum line length:
```js
const f = a => b => c => // could continue ad infinitum and never breaks
  find( a, b, c );
```
I think preventing a break for arrow function with a single simple argument is a reasonable change that shouldn't meet much opposition.

---
But formatting details aside, I think the main reason why this new attempt at Prettier is different is that it removes the need for the `eslint --fix` step. `wp-prettier` can insert all the whitespace inside parens correctly, as the WordPress convention requires, and doesn't need any other tool to fix it. There is a new `parenSpacing` option for that.

Everything else, as I see it, are just subtle differences in opinion about how things should be formatted, they are not specific to the "WordPress culture", if you know what I mean, and don't have a rational resolution that everyone can agree on.

If something seems obviously wrong, as the examples above, we can propose patches upstream. The only WordPress-specific difference seems to be the paren spacing.

ntwb commented:
Thanks for the detailed reply @jsnajdr 💯

I think it would great to see how this looks on the entire code base, a gist, a fork maybe, then the Gutenberg and WordPress teams can take a closer look...

I'm thinking once the above is available we can document what changes would be required and acceptable to update the WordPress Coding Standards with, and then create a P2 on make.w.org/core for further feedback

jsnajdr commented:
> I think it would great to see how this looks on the entire code base, a gist, a fork maybe, then the Gutenberg and WordPress teams can take a closer look...

I can reformat the whole codebase and push the diff to this branch. I originally decided to avoid that in order to keep the PR small.

gziolo commented:
> I can reformat the whole codebase and push the diff to this branch. I originally decided to avoid that in order to keep the PR small.

It would be fun to see a branch with all files formatted :)

jsnajdr commented:
I updated this PR with a commit that reformats all JS files in the codebase. 1200+ files, 18000+ lines :wink:

gziolo commented on package-lock.json:
```diff
@@ -7858,6 +7860,23 @@
 				"webpack-cli": "^3.1.2",
 				"webpack-livereload-plugin": "^2.2.0"
+			},
+			"dependencies": {
```
We will have to update how this gets resolved to avoid having `node_modules` inside `packages/scripts`. The easiest way to hoist it is to install the dependency as described in https://github.com/WordPress/gutenberg/tree/master/packages#adding-new-dependencies.

At this moment you can also install it in the root folder and remove afterwards.

gziolo commented on packages/scripts/scripts/format-js.js:
```diff
@@ -0,0 +1,76 @@
+// 2. use configuration provided by the project
+// 3. use the default provided in the scripts module
+let configArgs;
+const configFromCLI = getArgFromCLI( '--config' );
```
It looks like params passed to `prettier` are whitelisted. Interesting approach. In other places, we would pass whatever the users provide:
https://github.com/WordPress/gutenberg/blob/9783474bdd86df5b82137663bc2e7fec56c7d779/packages/scripts/scripts/lint-js.js#L19-L50

I'm just sharing for the full picture. It works for me the way it's implemented as well. It's more effort probably the way I see it here :)

jsnajdr replied:
I know I know, I spent some considerable time tuning the CLI args 🙂 The reason is that I can't forward all CLI arguments verbatim, because I need to change some of them.

Specifically, I need to convert the "file args" from directory names to globs (`packages/compose` to `packages/compose/**/*.js`) with the `dir-glob` package. And the CLI utils module doesn't provide helpers to separate the file args from regular args and forward only the regular ones.

The `lint-js` script always forwards all args without modification, it only adds some defaults when they are not present.

gziolo reviewed (commented):
@jsnajdr awesome work on this integration. I totally agree that having this kind of integration will help us focus on the logic rather than waste time on countless discussions on the code formatting ❤️ 

> Some ESLint rules that guard formatting are known to conflict with Prettier (difference in opinions about what's the "right" format), so I'm disabling these that check for things that are automatically handled by Prettier. There's also an `eslint-config-prettier` ESLint plugin for that.

Indeed, I see that there are some rules removed in:
https://github.com/WordPress/gutenberg/pull/18048/commits/ad9e952dc2caa33edf64ea0b817dde22a9d294ee
Should we move them to their own ruleset for those who would like to still use them rather than Prettier? Can you elaborate or visualize what works differently? I guess we will have to update JavaScript coding style guidelines to reflect that.

## ESLint warnings

There are only 92 errors reported by ESLint on Travis:
https://travis-ci.com/WordPress/gutenberg/jobs/248602039#L1105

Some of them are directly related to the ESLint disable rules which were in the same line. We probably should switch to the syntax which works with the next line or the one which surrounds code which is affected.

The most common violations are:
- ` 1792:28  error  Unexpected mix of '+' and '*'  no-mixed-operators`
- `63:30  error  A space is required after '{'   react/jsx-curly-spacing`

We should investigate them and if necessary relax them first, together with all required changes to ESLitn disable comments.

## Documentation

At the very last stage, this PR will have to include user documentation:
- example usage in `package.json` file in https://github.com/WordPress/gutenberg/tree/master/packages/scripts#setup
- its own section describing the new script, similar to `lint-js`: https://github.com/WordPress/gutenberg/tree/master/packages/scripts#lint-js
- a note in the CHANGELOG file, see https://github.com/WordPress/gutenberg/blob/master/packages/scripts/CHANGELOG.md#new-features

jsnajdr commented:
> Should we move them to their own ruleset for those who would like to still use them rather than Prettier?

I don't see what's the use case for that, but if it provides any value, why not?

> Can you elaborate or visualize what works differently? I guess we will have to update JavaScript coding style guidelines to reflect that.

The `operator-linebreak` rule (`:` should be placed at end of line) requires multiline ternaries to be formatted like this:
```js
'words' === settings.type ?
  matchWords( text, matchRegExp, settings ) :
  matchCharacters( text, matchRegExp, settings );
```
but Prettier prefers this:
```js
'words' === settings.type
  ? matchWords( text, matchRegExp, settings )
  : matchCharacters( text, matchRegExp, settings );
```

Then there is the `no-mixed-spaces-and-tabs` rule that doesn't like Prettier's formatting of indented parens inside ternaries:

<img width="312" alt="spaces-ternaries" src="https://user-images.githubusercontent.com/664258/67487137-b2043880-f66d-11e9-97af-6073e83edd66.png">

Do you see the two spaces (instead of a tab) before the closing `}`? That's to make the brace aligned to the opening `? `, which is always two characters no matter what the tab width is. Formatting this construct is a shady area where there's no perfect solution if tabs are used.

Then I see a lot of reports from the `indent` rule like:
```
Expected indentation of 7 tabs but found 8
```
but that's expected and obvious.

The `no-mixed-operators` rule wants parens in expressions like this: `a + ( b * c )`, but Prettier knows that the `*` operator has precedence anyway and doesn't add the parens: `a + b * c`.

The `react/jsx-curly-spacing` doesn't like `<div>{' '}</div>` and requires `<div>{ ' ' }</div>`. And it's totally right -- it's a bug in the fork that nobody else noticed yet!

It's a very special case where (note the space before `<hr/>` in the input)
```jsx
<div>{ a } <hr /></div>
```
when broken into multiple lines, needs to be formatted like:
```jsx
<div>
  { a }{ ' ' }
  <hr />
</div>
```
otherwise the JSX whitespace ignoring rules would ignore the space.

jsnajdr commented on package.json:
```diff
@@ -172,6 +172,7 @@
 		"fixtures:server-registered": "wp-scripts env docker-run php ./bin/get-server-blocks.php > test/integration/full-content/server-registered.json",
 		"fixtures:generate": "npm run fixtures:server-registered && cross-env GENERATE_MISSING_FIXTURES=y npm run test-unit",
 		"fixtures:regenerate": "npm run fixtures:clean && npm run fixtures:generate",
+		"format-js": "wp-scripts format-js",
```
Yes, it would be nice to add a formatting pre-commit step with `lint-staged`. I didn't add that to the original PR, but can add it now if we agree about the desired workflow.

jsnajdr commented on .prettierrc.js:
```diff
@@ -0,0 +1,3 @@
+// Import the default config file and expose it in the project root.
```
I'm not familiar with all IDE integrations, but the Atom one I use works as follows:
- tries to find `prettier` in the project's root `node_modules`. The fork must be there, having it in, say, `packages/wp-scripts/node_modules/prettier` or `node_modules/@wordpress/scripts/node_modules` is not enough. If the fork is not found, it uses the default bundled Prettier, which doesn't have the `--paren-spacing` option
- runs Prettier with cwd set as project root and using the default config-searching algorithm, making it mandatory for the config to be in the root. One can probably configure the IDE integration to look for the Prettier config somewhere else, but that's an extra manual step everyone would be required to do.

jsnajdr commented:
I rebased this onto the latest `master`. Ready for review and comment (or maybe for approval and merge) again.

ntwb commented:
Before this is merged a post should be created for the Make-Core P2 proposing the switch to Prettier, as this makes significant changes to the WordPress Coding Standards these changes need be be discussed and time allowed for the WordPress Core Team to review the changes.

youknowriad commented:
> this makes significant changes to the WordPress Coding Standards 

AFAIK, I don't think this makes changes to the standards but it enforces new things to be consistent (line height...) that might not be standard today.

A make/core post would be great anyway. Any volunteer?

gziolo commented:
This PR was discussed during the last Core JS weekly chat. The full summary is available at:
https://make.wordpress.org/core/2019/11/12/javascript-chat-summary-november-12-2019/

The related part is:

> Some discussion ensued on what is needed to merge this pull.
> 
>-  Post on make.wordpress.org regarding any code standard changes affected by this pull (and to inform about the change). This will need further feedback from core lead devs, core committers and contributors.
> - There are some rule removals from the WordPress ESLint plugin package, some discussion ensued about how to handle that and are those changes necessary. At a minimum, there seems to be some agreement that the ESLint rules should match the standards.
>
> Tasks:
> 
> - Check how many of the existing eslint rules could stay unmodified and what can be tweaked to work with Prettier and which need to be disabled. Should align with whatever standards are for the project (including any necessary modifications to those standards)
> - We should also seek a way so that the eslint plugin can continue to be used without prettier so folks can have the benefit of styling standard checks the same way as of today.
> - Once the above is done, write a post on make.wordpress.org/core outlining the changes and inviting feedback.

jsnajdr commented:
**Differences with [WordPress JavaScript Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/)**

I found three places where Prettier does something different from the recommendations and where the standard will need to updated:

**Formatting ternaries and binary ops:** The standard [says](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#multi-line-statements) that when breaking a long statement, the "operator" should be at the end of line, and doesn't distinguish between binary and ternary operators. But Prettier does. When breaking a binary operator, it indeed puts the operator at the end of line:
```js
const result =
  partOne +
  PartTwo;
```
But the ternary operators are put on the start of the new line (after the indentation):
```js
const result = isRtl
  ? 'right'
  : 'left';
```

**Wrapping chained n-ary operators:** Another difference related to binary operators and the [Multiline statements section](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#multi-line-statements) is that Prettier puts each operand on separate line, even the left side of an assignment. It doesn't do the "fluid text wrap" style like this:
```js
const result = a + b +
  c + d;
```
but this:
```js
const result =
  a +
  b +
  c +
  d;
```

**Chained method calls and context:** The standard [recommends](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#chained-method-calls) that when one method in a chain "changes the context", it should add an extra indentation:
```js
elements
  .addClass( 'foo' )
  .children()
  .html( 'hello' )
  .end()
  .appendTo( 'body' );
```
But Prettier doesn't know what the context is -- it's an ambiguous concept even for a human. So it indents everything equally:
```js
elements
  .addClass( 'foo' )
  .children()
  .html( 'hello' )
  .end()
  .appendTo( 'body' );
```

It also adds one extra touch: if the initial identifier is just one or two characters long, it will keep the first method call on the same line:
```js
el.addClass( 'foo' )
  .children()
  .html( 'hello' )
  .end()
  .appendTo( 'body' );
```

jsnajdr commented:
**Differences from the current ESLint rules config**
ESLint formatting rules that Gutenberg currently uses and Prettier often have different opinions about what's the right formatting. The Prettier authors don't recognize ESLint as an authority that decides what's right or wrong, but establish their own rules and conventions.

That's why ESLint (more precisely, it's formatting-only rules) and Prettier cannot live along and why keeping the ESLint config for these who prefer it is not tenable.

Details of what the differences are:

**`operator-linebreak`**
The rule (`:` should be placed at end of line) requires multiline ternaries to be formatted like this:
```js
'words' === settings.type ?
  matchWords( text, matchRegExp, settings ) :
  matchCharacters( text, matchRegExp, settings );
```
but Prettier prefers this:
```js
'words' === settings.type
  ? matchWords( text, matchRegExp, settings )
  : matchCharacters( text, matchRegExp, settings );
```

**`no-mixed-spaces-and-tabs`**
This rule that doesn't like Prettier's formatting of indented parens inside ternaries:

<img width="312" alt="spaces-ternaries" src="https://user-images.githubusercontent.com/664258/67487137-b2043880-f66d-11e9-97af-6073e83edd66.png">

Do you see the two spaces (instead of a tab) before the closing `}`? That's to make the brace aligned to the opening `? `, which is always two characters no matter what the tab width is. Formatting this construct is a shady area where there's no perfect solution if tabs are used.

The WP coding standard is OK with these mixed tabs and spaces. It [says](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#spacing):

> Spaces may align code within documentation blocks or within a line, but only tabs should be used at the start of a line.

**`indent`**
This rule often reports warnings like:
```
Expected indentation of 7 tabs but found 8
```
There are many different reasons why exactly a certain piece of code is formatted differently -- `indent` and Prettier have different line-breaking and indenting algorithms.

**`no-mixed-operators`**
This rule wants parens in expressions like this: `a + ( b * c )`, but Prettier knows that the `*` operator has precedence anyway and doesn't add the parens: `a + b * c`.

jsnajdr committed:
Add Prettier NPM dependency to packages/scripts and top-level

jsnajdr committed:
Scripts: Add Prettier formatting script

jsnajdr commented:
Some updates on this PR:
- I upgraded the `wp-prettier` package to version 1.19.1 that was released earlier this week. Besides picking up changes from the upstream version, it fixes the `{' '}` vs `{ ' ' }` JSX formatting bug found by @gziolo.
- added `prettier` dependency also to the top-level `package.json`, so that we ensure that the correct version is installed in the top-level `node_modules` and is picked up by the formatting scripts and IDEs. If we don't do that, npm will dedupe the version required by Storybook into the root. Ugly but effective.
- removed the actual formatting from this PR, so that it's smaller and easier to review, and also to avoid it being in conflict with `master` all the time. Anyone can run the formatting themselves by running `npm run format-js`
- posted some comments about how Prettier relates to WordPress JS Coding Standards and ESLint's formatting rules. Addresses concerns raised during this week's Core JS Slack chat.

gziolo commented:
> * added `prettier` dependency also to the top-level `package.json`, so that we ensure that the correct version is installed in the top-level `node_modules` and is picked up by the formatting scripts and IDEs. If we don't do that, npm will dedupe the version required by Storybook into the root. Ugly but effective.

This is probably something that we should include in the `README.md` of `@wordpress/scripts`. Maybe troubleshooting section or something like that? I guess some hints how to setup everything with IDEs would be a good follow-up (I don't think its unique for Prettier).

In addition, we really need to have `format-js` command documented in the first place so people could start using it :)

gziolo commented on packages/scripts/package.json:
```diff
@@ -44,13 +44,15 @@
 		"lodash": "^4.17.15",
 		"minimist": "^1.2.0",
 		"npm-package-json-lint": "^4.0.3",
+		"prettier": "npm:wp-prettier@^1.19.1",
```
This remind me that we need to add `engine` section and set minimum version of `npm` to 6.9.

In addition, we need to update the defaults in `check-engines` command and probably mention then now in docs given that they are quite specific:

https://github.com/WordPress/gutenberg/blob/ca0c0eb6ad8508154169b89d03c119c0a56dad0b/packages/scripts/scripts/check-engines.js#L24

https://github.com/WordPress/gutenberg/tree/master/packages/scripts#check-engines

Maybe a subsection in `Advanced Usage` on how to update `npm` to the latest version would be nice to have as well:

https://github.com/WordPress/gutenberg/tree/master/packages/scripts#advanced-usage

aduth commented:
> The rule (`:` should be placed at end of line) requires multiline ternaries to be formatted like this:

I think this can be configured in ESLint specific to the ternary operators:

https://eslint.org/docs/rules/operator-linebreak#options

aduth commented:
> **`no-mixed-operators`**
> This rule wants parens in expressions like this: `a + ( b * c )`, but Prettier knows that the `*` operator has precedence anyway and doesn't add the parens: `a + b * c`.

This seems so strange to me, as it optimizes the pretty output for the machine, not the human. I've always seen this as a rule to exist purely for the sake of making the reader's life easier, to not have to consider order of operations.

It would not be a deal-breaker for me, but definitely seems to yield a worse developer experience.

jsnajdr commented:
> This seems so strange to me, as it optimizes the pretty output for the machine, not the human.

@aduth I don't disagree. Prettier adds extra parens in many instances where the precedence and associativity rules would yield the same results:
- `a / b * c` is the same as `( a / b ) * c`
- `a === b === c` is the same as `( a === b ) === c`
- `a + b % c` is the same as `a + ( b % c )`

It seems it's [assumed that "everyone knows"](https://github.com/prettier/prettier/issues/187#issuecomment-306769450) that multiplication and divison have precedence over subtraction and addition -- unlike the other rules, it's an elementary school material.

The exact rules can be seen in the [needsParens](https://github.com/prettier/prettier/blob/bf2989718b83ff81f994b5a663833a1d239fb76c/src/language-js/needs-parens.js#L342-L386) and [shouldFlatten](https://github.com/prettier/prettier/blob/bf2989718b83ff81f994b5a663833a1d239fb76c/src/common/util.js#L387-L427) functions.

jsnajdr commented:
> I think this can be configured in ESLint specific to the ternary operators:

That's true. The default configuration from ESLint is Prettier-compliant:
```
"after", { "overrides": { "?": "before", ":": "before" } }
```
but the Gutenberg ESLint config overrides it to just `"after"`, with no special treatment for the ternary operation.

This ESLint rule will need to be updated to conform with the updated coding standard.

jsnajdr committed:
ESLint config: use default (linebreak before ternary op) config for operator-linebreak rule

jsnajdr committed:
ESLint: disable formatting rules in the recommended config

Adds `eslint-config-prettier` to the recommended config and creates an alternative
`recommended-with-formatting` config for project that want to keep ESLint formatting
rules and opt-out of Prettier.

jsnajdr commented:
Pushed several changes:
- don't remove the formatting rules from the ESLint config. Rather disable them by including [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) in the `recommended` config. They are still available as `recommended-with-formatting` config for these who want to opt out (requested by @aduth) among others.
- update the configuration of the `operator-linebreak` rule to require `?` and `:` operators (ternary) to be after linebreak rather than before. Unlike binary operators like `+`. Done by using the default supplied by ESLint core and removing our override.

The result is that by default, recommended ESLint configs provided by the `@wordpress/*` toolset don't check formatting and a Prettier script and config are provided. But projects can opt-out by not using the `format-js` script and by using the `recommended-with-formatting` ESLint config instead. How does that sound?

ntwb commented:
A thought just crossed my mind whilst working on the VS Code stylelint extension, I noticed these VS Code settings for Prettier:

```json"
  "prettier.arrowParens": "always",
  "prettier.bracketSpacing": true,
  "prettier.jsxBracketSameLine": false,
  "prettier.printWidth": 100,
  "prettier.semi": true,
  "prettier.singleQuote": true,
  "prettier.tabWidth": 2,
  "prettier.trailingComma": "es5",
  "prettier.useTabs": true,
```

How would would the `parenSpacing: true,` setting work with the VS Code extension?

I suspect the VS Code Prettier extension will ignore the `"prettier.parenSpacing": true,` setting if it was added...

Will need to do a bit more research on this to determine how the extension detects configs, Prettier uses _cosmiconfig_, though I'm unsure what configuration would take precedence, the config from the repo, or the VS Code extension setting.

@noisysocks I know you've been using VS Code and Prettier forever with Gutenberg, have you hit upon this issue or have some thoughts they would be greatly appreciated...

jsnajdr commented:
@ntwb I don't know how the VSCode Prettier integration works, but I'm familiar with the Atom integration, which works as follows:

**Configuration:** The IDE integration tries to find a Prettier config file (`.prettierrc` or one of the many alternative names, including a `package.json` field) and uses that when formatting files inside the project. The configuration in the IDE settings is only a fallback, default that is used if no other config files is found.

**Prettier binary:** Similarly, the Atom integration tries to find a Prettier binary in the project's `node_modules` folder and will use it if found. Only when it's not found will it use the Prettier that is bundled and shipped together with the plugin. If Prettier is in `node_modules`, it can be ensured that all contributors use the same version. Also, in our case, where we use a fork, the difference from the IDE-bundled Prettier is much bigger.

For both of the above to work correctly, the Prettier config needs to be in the project's top-level folder and the Prettier binary needs to be in the top-level `node_modules` folder. If it's in any other location, e.g., in a `packages/scripts` subfolder, IDEs and other tools won't find it reliably.

If the project-wide config and binary are set up correctly, the defaults of the IDE integration should not matter.

This is true for the Atom integration I'm familiar with, but I'd expect that other editor integration work exactly the same way.

jsnajdr committed:
Scripts: the format-js script checks if Prettier is present and has top-level config

Prettier presence in a `wp-scripts`-powered project is optional, and the `format-js` script
checks if it's there and if it's indeed the fork (`wp-prettier`). Will report error otherwise.

Also, require a top-level Prettier config to be present. We can't default to the one inside
`wp-scripts`, because IDE and editor integrations won't find it there and will use the Prettier
defaults.

jsnajdr commented:
Latest updates for the `format-js` script in `wp-scripts`:

The script requires that the Prettier fork is installed in the project root and checks that it's indeed the `wp-prettier` fork and not the official `prettier`. It will report an error if Prettier is not found or unexpected version.

Prettier is no longer a dependency of the `wp-scripts` package itself. For IDE and editor integrations to work reliably, it always needs to be installed in the project's root `node_modules`. Not somewhere in `node_modules/@wordpress/scripts/node_modules/prettier` -- the editor integrations won't find it there.

For the same reason, we require that a project has a top-level `.prettierrc.js` file. The `format-js` script doesn't default to the one that ships inside `packages/wp-scripts/config`. Again, editor integrations wouldn't find it there.

A project can add an one-liner `.prettierrc.js` config file and everything will be fine:
```js
module.exports = require( '@wordpress/scripts/config/.prettierrc.js' );
```

ntwb commented:
Thanks for the detailed report @jsnajdr, I'll need to test VS Code but I would suspect it would perform in a similar fashion.

Thanks for updating the PR to account for this, and the docs tweak https://github.com/WordPress/gutenberg/pull/19074#discussion_r358330696

We will need to update the `@wordpress/scripts` later to also make sure this is well documented.

jsnajdr commented:
There's an e2e failure in this PR (even after a rebase) that seems 100% unrelated to Prettier:
```
FAIL packages/e2e-tests/specs/editor/blocks/buttons.test.js (7.133s)
  Buttons
    ✓ has focus on button content (3195ms)
    ✕ can jump to the link editor using the keyboard shortcut (3342ms)
  ● Buttons › can jump to the link editor using the keyboard shortcut
    expect(received).toMatchSnapshot()
    Snapshot name: `Buttons can jump to the link editor using the keyboard shortcut 1`
    - Snapshot
    + Received
      "<!-- wp:buttons -->
      <div class="wp-block-buttons"><!-- wp:button -->
    - <div class="wp-block-button"><a class="wp-block-button__link">WordPress</a></div>
    + <div class="wp-block-button"><a class="wp-block-button__link" href="https://wwww.wordpress.o" title="https://wwww.wordpress.o">WordPress</a></div>
      <!-- /wp:button --></div>
      <!-- /wp:buttons -->"
      28 | 		await page.keyboard.press( 'Enter' );
      29 | 
    > 30 | 		expect( await getEditedPostContent() ).toMatchSnapshot();
         | 		                                       ^
      31 | 	} );
      32 | } );
      33 | 
      at Object.toMatchSnapshot (specs/editor/blocks/buttons.test.js:30:42)
          at runMicrotasks (<anonymous>)
 › 1 snapshot failed.
```
@jorgefilipecosta any idea what's wrong here? Also cc @adamsilverstein who asked about the failure in #core-js Slack.

aduth commented:
@jsnajdr See also #19490. It's a known intermittent failure. I'll restart the build. I'm hopeful we can get the underlying issue resolved soon.

jsnajdr commented:
> It's a known intermittent failure. I'll restart the build.

Thanks! I got the failure two times in a row and that's what confused me 🙂

jsnajdr commented:
@gziolo @aduth this PR still needs review. I'm not aware of any further issues that would need to be resolved. Can you help me with getting it shipped?

gziolo commented on packages/scripts/scripts/format-js.js:
```diff
@@ -0,0 +1,114 @@
+		hasPackageProp( 'prettier' );
+
+	if ( ! hasProjectPrettierConfig ) {
+		return {
```
It's less pressing for the start but I would like to find the right path moving forward.


I find this part controversial given how all other scripts work. It should work perfectly fine without the config file defined in the root of the project. This is how it is handled elsewhere. I understand that it makes it tricky to work with IDEs but it gives the possibility to run it from CLI on demand, so it still allows some automation.

I'd love to see some hybrid approach (probably applicable also for linters and test runners). When the script runs for the first time, it informs that the IDE integration is missing and asks whether to set up one. If the answer is `yes` then the config is created in the roof of the project. What do you think?

gziolo commented on packages/scripts/scripts/format-js.js:
```diff
@@ -0,0 +1,76 @@
+}
+
+// Converts `foo/bar` directory to `foo/bar/**/*.js`
+const globArgs = dirGlob( fileArgs, { extensions: [ 'js' ] } );
```
Nice, I like what you did here. It's a great idea 👍

gziolo commented on .prettierrc.js:
```diff
@@ -0,0 +1,3 @@
+// Import the default config file and expose it in the project root.
+// Useful for editor integrations.
+module.exports = require( '@wordpress/scripts/config/.prettierrc.js' );
```
We should document it and ensure that this file is always usable outside. I guess it will become the blessed way to integrated with IDEs. Unless we expose a function that proxies the default configs provided.

```js
const { getDefaultConfig } from '@wordpress/scripts';

module.exports = getDefaultConfig( 'prettier' );
```

It's probably unnecessary indirection though :)

gziolo reviewed (commented):
There are two tasks left and we are good to go:

- we need to document changes introduced in CHANGELOG files for the ESLint plugin and `@wordpress/scripts` package
- we need to update related docs in both packages update
  - there needs to be a section on how to run `format` command with `wp-scripts`, how to integrate with IDE
  - for the ESLint plugin we need to add a note about `recommended-with-formatting` ruleset which gets exposed

gziolo reviewed (approved):
I tested it locally again, it works like a charm. Gutenberg still works after all the refactorings applied :)

Let's get it in and work on the listed items in parallel to the PR which will format the whole codebase 💥

gziolo commented:
I almost forgot, which version of npm is required to use aliases? I think we should enforce it in Gutenberg at least.

jsnajdr commented:
> We would need to enforce npm v6.9.0+

Yes, it was also mentioned in this comment: https://github.com/WordPress/gutenberg/pull/18048#discussion_r337402830

ntwb commented on packages/eslint-plugin/configs/recommended.js:
```diff
@@ -1,39 +1,3 @@
-			],
-		},
-	],
+	extends: [ require.resolve( './recommended-with-formatting.js' ), 'eslint-config-prettier' ],
```
Apologies for the lateness of this here, by removing the `recommended` config here and switching to `recommended-with-formatting` this may be troublesome for projects currently consuming `@wordpress/scripts` and using the ESLint config.

Will this affect those users in that the ESLint config is no longer available and they are not yet ready to switch to or know anything about Prettier?

p.s. I've not had the ability to study this use case in depth

jsnajdr replied:
@ntwb The `recommended` config is still available. This is how it works after this patch:
- the `recommended-with-formatting` config includes all the rules that check code formatting, and that sometimes conflict with Prettier conventions. It's basically the former `recommended` config with a new name. It's intended for projects that want to opt-out from Prettier and continue to use ESLint for formatting.
- the `recommended` config continues to exist, and it extends the `recommended-with-formatting` config, and disables all the code formatting rules by applying `eslint-config-prettier`. Therefore, the new default from now is to _not_ check code formatting with ESLint, and use solely Prettier for code formatting.

> Will this affect those users in that the ESLint config is no longer available and they are not yet ready to switch to or know anything about Prettier?

If you just upgrade to the new version of `@wordpress/scripts` and do nothing else, ESLint will stop reporting code formatting issues. Until you add Prettier to your project explicitly, not tool will be handing code formatting.

ntwb replied:
These changes should also be noted in the changelog

I'm still not sure on the idea of removing the ability for users to be able to format their code, need to think on this some more, I'm fully behind the switch to Prettier so by not doing this it may help in this regard and nudge users to add Prettier once they discover their code is no longer formatted 😏

----

Also, just looking at the code of the two files, instead of creating the _new_ `eslint-plugin/configs/recommended.js` config file, why not leave the existing `eslint-plugin/configs/recommended-with-formatting.js` fille as is and add `'eslint-config-prettier'` to `extends` in `recommended.js`?

There are no other changes to the overall ESLint config besides this one, and if it is to be the default anyway, thoughts?

e.g.:

```js
	extends: [
		'eslint-config-prettier'
		require.resolve( './jsx-a11y.js' ),
		require.resolve( './custom.js' ),
		require.resolve( './react.js' ),
		require.resolve( './esnext.js' ),
	],
```

jsnajdr replied:
> Also, just looking at the code of the two files, instead of creating the _new_ `eslint-plugin/configs/recommended-with-formatting.js` config file, why not leave the existing `eslint-plugin/configs/recommended.js` fille as is and add `'eslint-config-prettier'` to `extends` in `recommended.js`?

@ntwb I'm not sure I understand your comment -- I believe this is exactly what I'm doing in this PR. Rename `recommended.js` to `recommended-with-formatting.js` with the content not changed, and then create a _new_ `recommended.js` that extends the `-with-formatting` one. The Git diff doesn't show the change as rename, but it's what happened.

jsnajdr replied:
> These changes should also be noted in the changelog

Added an entry to `packages/eslint-plugin/CHANGELOG.md`

ntwb replied:
Thanks, let me rephrase it, why create the new `recommended-with-formatting.js` file?

It's not required, the only change required is adding `eslint-config-prettier` to `extends` in the original `recommended.js` file.

I have created this PR against this branch demonstrating the using ESLint `--print-config`

> https://github.com/WordPress/gutenberg/pull/19744

jsnajdr replied:
> why create the new `recommended-with-formatting.js` file?

Oh, I see the point now. In the November 19th core-js chat (Slack link [here](https://wordpress.slack.com/archives/C5UNMSU4R/p1574173838080600), no blog post summary found for that week), we discussed a proposal by @aduth to support two ESLint modes:
- one doesn't run the code formatting rules and let's Prettier completely take care of that. That's the new `recommended` behavior.
- the second still runs the code formatting roles, as it did until now, for projects that for some reason want to opt out from Prettier and still use `wp-scripts`. That's the `recommended-with-formatting` config. It intentionally provides an escape hatch to use the old convention.

ntwb replied:
Right, in that case `eslint-config-prettier` should be in the `recommended-with-formatting.js` file, not the `recommended.js` file because as it stands both configurations have identical configurations at the moment.

gziolo replied:
It's all good as is according to what we agreed during the core-js chat. `recommended-with-formatting` is what we had so far where ESLint does some code styling checks, whereas `recommended` has many of the rules disables and lets Prettier own it with formatting.

jsnajdr replied:
> I think I have now fully confused myself, I'll come back to this later today

Yes, there are at least two facts that are indeed confusing:
- `recommended-with-formatting` is used for projects _without_ Prettier formatting (but _with_ ESLint formatting checks)
- `eslint-config-prettier` disables existing rules rather than adding any new ones

ntwb replied:
Thanks, and sorry

gziolo commented:
Related discussion on WordPress Slack in #core-js channel (link requires registration): https://wordpress.slack.com/archives/C5UNMSU4R/p1579085530105700

let’s make sure all WordPress build servers have npm v6.9 or later installed before we land Prettier changes so we accidentally don't break the workflow.

gziolo committed:
Bump the minimum required version of npm to 6.9.0

gziolo commented:
I added commit which enforces `npm` v6.9.0 or later: https://github.com/WordPress/gutenberg/pull/18048/commits/b6b510961ef74a5ce693cdcf47b1d5c3c176960d. I also marked it as a breaking change for `@wordpress/scripts`.

jsnajdr committed:
Add ESLint config changes notice to changelog

ntwb reviewed (dismissed):
The new ESLint config file is not required.

The only ESLint config change required is the addition of `eslint-config-prettier` to the original `recommended.js` file

ntwb commented:
### Note: 

I've created https://meta.trac.wordpress.org/ticket/4974 to have the w.org "build server" upgraded, merging this PR or #19680 and merging to w.org SVN may cause explosions before this upgrade occurs

chrisvanpatten commented:
Very excited to see this land — while I'm not super familiar with prettier yet, having recently spent several days (!) configuring eslint, it's really cool to see an alternative approach and I'm looking forward to experimenting with it.

In addition to the internal documentation needs that @gziolo mentioned previously, it would be great to see this get the "Needs Dev Note" label and have a dev note published to make/core explaining a few things in simple and direct language:

1. Why Prettier? What are the benefits it will bring to developers?
2. How would one use it via the `@wordpress/scripts` package?
3. What is the "venn diagram" of Prettier and ESLint? In other words, how do the two tools solve different problems (or not, as the case may be)?
4. What would the upgrade path be for the average developer leveraging `@wordpress/scripts`?

I know some of these items are hinted at in the [proposal devnote](https://make.wordpress.org/core/2019/12/09/proposed-javascript-coding-standards-revisions-for-prettier-compatibility/), but once the changes land and are available to third parties, it'd be great to have these higher level items clearly documented.

Now to be fair that may not need to happen _immediately_, as some of these answers may not be obvious yet, but I think it will help those of us using `wp-scripts` as the basis of our linting and build processes understand the scope of the change and how to take advantage.

mkaz commented:
@chrisvanpatten Documentation for prettier was added in #19074  and available at: https://developer.wordpress.org/block-editor/contributors/develop/getting-started/#prettier

Any suggestions for where else you think more might be needed?

chrisvanpatten commented:
@mkaz That page seems specific to usage of Prettier _within Gutenberg_, but it seems like it will also be available to third parties via `wp-scripts`. I haven't seen much about that external usage — how the transition works, where you would want to use Prettier vs ESLint, what the commands are to use it, etc.

ntwb commented on packages/scripts/README.md:
````diff
@@ -16,6 +16,8 @@ You only need to install one npm module:
 npm install @wordpress/scripts --save-dev
 ```
 
+**Note**: This package requires `npm` 6.9.0 or later, and is not compatible with older versions.
````
Should this also mention Node.js v10 also?

It is noted as required in the changelog here https://github.com/WordPress/gutenberg/pull/18048/files#diff-748355c268adfa7ea2255a3045d81ed9R5

gziolo replied:
Yes, we can extend it.

mkaz commented:
I didn't want to disrupt this PR with more commits so added readme docs in #19766 

You can either cherry-pick or merge after if it covers.

cc: @chrisvanpatten

jsnajdr committed:
Update package-lock.json

gziolo committed:
Regenerate package-lock.json

gziolo commented on packages/scripts/package.json:
```diff
@@ -53,6 +55,7 @@
 		"lodash": "^4.17.15",
 		"minimist": "^1.2.0",
 		"npm-package-json-lint": "^4.0.3",
+		"prettier": "npm:wp-prettier@1.19.1",
```
I added `prettier` as a dependency here as well. I hope it is a good idea.

ntwb replied:
Untested, but I "think" this is correct

gziolo commented:
@jsnajdr, I regenerated `package-lock.json` using the latest `master` version in 
d19a844. I think we also need to define `prettier` as a dependency of `@wordpress/scripts` to make it work smoothly. Feel free to merge if you agree with the changes applied.

jsnajdr commented:
> I think we also need to define `prettier` as a dependency of `@wordpress/scripts` to make it work smoothly. Feel free to merge if you agree with the changes applied.

Thanks @gziolo! I think that Prettier doesn't necessarily need to be a dependency of `@wordpress/scripts`, but let's resolve this in followup discussions about the IDE integration and config files.

Now, let's merge this PR before it gets stale again! 🙂

gziolo commented:
Awesome work @jsnajdr, thanks for bringing Prettier to the WordPress community 🎉

ntwb commented:
Woot 🥳 

Thanks to everyone who contributed along the way, your feedback and patience is greatly appreciated 💯

Special shoutout and thanks @jsnajdr for maintaining, persisting, and persevering for the longest time to bring this to the WordPress project 🙇🏼‍♂️

gziolo commented:
I opened https://github.com/WordPress/gutenberg/pull/19946 that document this new script added to `@wordpress/scripts`. @jsnajdr, can you double-check the proposal? Feel free to add your own commits if you feel like it needs more details. I omitted the part about IDEs as it feels like we need to experiment with it a bit first.

epiqueras commented on packages/eslint-plugin/configs/recommended.js:
```diff
@@ -1,39 +1,3 @@
-			],
-		},
-	],
+	extends: [ require.resolve( './recommended-with-formatting.js' ), 'eslint-config-prettier' ],
```
We need to run prettier on lint if we are not going to enforce the formatting rules through eslint anymore.
